test/collection_test.js

```diff
@@ -0,0 +1,653 @@
+            sinon.match(v => v.length === 1 && v[0].title === "foo"),
+            {},
+            {safe: true});
+        });
+    });
+
+    it("should update published records local status", () => {
+      var batch = sandbox.stub(articles.api, "batch").returns(Promise.resolve({
+        published: [records[0]]
+      }));
+      return articles.pushChanges(result)
+        .then(res => res.published)
+        .should.eventually.become([
+          {
+            _status: "synced",
+            id: records[0].id,
+            title: "foo",
+          }
+        ]);
+    });
+
+    it("should delete unsynced virtually deleted local records", () => {
+      return articles.delete(records[0].id)
+        .then(_ => articles.pushChanges(result))
+        .then(_ => articles.get(records[0].id, {includeDeleted: true}))
+        .should.be.eventually.rejectedWith(Error, /not found/);
+    });
+
+    it("should locally delete remotely deleted records", () => {
+      var batch = sandbox.stub(articles.api, "batch").returns(Promise.resolve({
+        published: [Object.assign({}, records[1], {deleted: true})]
+      }));
+      return articles.pushChanges(result)
+        .then(res => res.published)
+        .should.eventually.become([
+          {
+            id: records[1].id,
+            deleted: true
+          }
+        ]);
+    });
+  });
+
+  describe("#sync", () => {
+    const fixtures = [
+      {title: "art1"},
+      {title: "art2"},
+      {title: "art3"},
+    ];
+    var articles;
+
+    beforeEach(() => {
+      articles = testCollection();
+      sandbox.stub(api, "batch").returns({
+        errors:    [],
+        published: [],
+        conflicts: [],
+      });
+      return Promise.all(fixtures.map(fixture => articles.create(fixture)));
+    });
+
+    it("should load fixtures", () => {
+      return articles.list()
+        .then(res => res.data)
+        .should.eventually.have.length.of(3);
+    });
+
+    it("should fetch latest changes from the server", () => {
+      var fetchChangesSince = sandbox.stub(articles.api, "fetchChangesSince")
+        .returns(Promise.resolve({
+          lastModified: 42,
+          changes: []
+        }));
+      return articles.sync().then(res => {
+        sinon.assert.calledTwice(fetchChangesSince);
+      });
+    });
+
+    it("should store latest lastModified value", () => {
+      var fetchChangesSince = sandbox.stub(articles.api, "fetchChangesSince")
+        .returns(Promise.resolve({
+          lastModified: 42,
+          changes: []
+        }));
+      return articles.sync().then(res => {
+        expect(articles.lastModified).eql(42);
+      });
+    });
+
+    it("should resolve early on pull failure", () => {
+      const result = new SyncResultObject();
+      result.add("conflicts", [1]);
+      sandbox.stub(articles, "pullChanges").returns(Promise.resolve(result));
+      return articles.sync()
+        .should.eventually.become(result);
+    });
+  });
```

Comment:
I don't know how we could test that, but that would be nice to have a test that specifies the order of the operations (pull / push / pull)